### PR TITLE
feat: use Discord rich embed format for webhooks

### LINKED ISSUE #105: Add the option for embedded notifications
I use this plugin because it's by far the most complete out of all and imo the least buggy.

But one thing I would like to get added is the option for embedded notifications. In my opinion they are better to read and look less cluttered than non embedded messages. I use this plugin for every notification in the game possible, instead for loot. For that I use an other plugin so it looks like this:

image

Is it possible for Dink to also have an option for embedded notifications? I know this is a personal preference and probably for a lot a people not a must-have, but I wanted to know what u thought about it.

Happy holidays!

### PR DESCRIPTION
**Pull request by iProdigy**
Closes #105

### CODE REVIEW

File: src/main/java/dinkplugin/message/NotificationBody.java [hunk 1]
```diff
@@ -1,39 +1,40 @@
...
+     */
     @Builder.Default
     @EqualsAndHashCode.Exclude
     private transient String screenshotFile = "image.png";
+    private NotificationType type;
+    private String playerName;
+    private T extra;
+
+    @SerializedName("text") // used by slack; avoids duplication in discord webhook
+    private String content;
```
Comment (iProdigy) [inline src/main/java/dinkplugin/message/NotificationBody.java:30]:
```suggestion
    private transient String content;
```
Alternatively, we could exclude this text from the json altogether (via `transient`) - I'm open to opinions

File: src/main/java/dinkplugin/message/NotificationType.java [hunk 2]
```diff
@@ -1,16 +1,24 @@
...
+    PET("Pet Obtained", "https://oldschool.runescape.wiki/images/Pet_list_icon.png"),
+    QUEST("Quest Completed", "https://oldschool.runescape.wiki/images/Quests.png"),
+    SLAYER("Slayer Task", "https://oldschool.runescape.wiki/images/Slayer_icon_%28detail%29.png"),
+    SPEEDRUN("Quest Speedrunning", "https://oldschool.runescape.wiki/images/Quest_Speedrunning_logo.png"),
+    KILL_COUNT("Completion Count", "https://oldschool.runescape.wiki/images/Enchanted_gem_detail.png"),
+    COMBAT_ACHIEVEMENT("Combat Achievement", "https://oldschool.runescape.wiki/images/Combat_achievements_detail.png"),
+    ACHIEVEMENT_DIARY("Achievement Diary", "https://oldschool.runescape.wiki/images/Achievement_Diaries.png");
+
+    private final String title;
+    private final String thumbnail;
```
Comment (iProdigy) [inline src/main/java/dinkplugin/message/NotificationType.java:23]: some of the images may be too small (e.g., pet) or too large (e.g., speedrun)...

File: src/main/java/dinkplugin/message/DiscordMessageHandler.java [hunk 3]
```diff
@@ -145,4 +156,21 @@ public void onResponse(@NotNull Call call, @NotNull Response response) {
             }
         });
     }
+
+    private static void injectContent(@NotNull NotificationBody<?> body, boolean screenshot) {
+        List<Embed> embeds = new ArrayList<>(body.getEmbeds() != null ? body.getEmbeds() : Collections.emptyList());
+        embeds.add(0,
+            Embed.builder()
+                .author(new Author(body.getPlayerName()))
+                .color(Utils.PINK)
```
Comment (iProdigy) [inline src/main/java/dinkplugin/message/DiscordMessageHandler.java:165]: Should we do a different color per notifier type instead?

**Comment by Felanbird** [comment]
Initial Thoughts:

I don't dislike the small icons, I think the large icons are generally more intrusive than they need to be.
Specifically, a collection log notif will never be a size where a large notif fits, while a clue notif will never not be huge, even if you get 1 item while opening on entrana.
image
image


I went ahead and tested 2 additional options for clue notifs, one being the old icon which was the same color for all variations, but it only has a large variation.
image

The other option was respecting what type of clue scroll it is, and using the image based on that. 
`I used hard icon with current handling for the testing` 
image

File: src/main/java/dinkplugin/message/NotificationBody.java (hunk 1 above)
Comment (pajlada) [reply src/main/java/dinkplugin/message/NotificationBody.java:30]: Unless we have a specific request for this, I don't think we should include it

File: src/main/java/dinkplugin/message/DiscordMessageHandler.java (hunk 3 above)
Comment (pajlada) [reply src/main/java/dinkplugin/message/DiscordMessageHandler.java:165]: I don't know if I could connect a color to a notifier (i.e. loot is yellow, combat achievments are green) so we'd probably end up picking them at random which feels off so unless someone has a good scheme for this I'd say just keep it single-color for now.

**Comment by pajlada** [comment]
Question: We're still sending it according to Discord's webhook format, we're just formatting the message differently into Discord's [rich embedded objects](https://discord.com/developers/docs/resources/channel#embed-object) rather than sticking everything into the "content" field right?

Could we have some examples of full payloads before and after?

**Comment by iProdigy** [comment]
Yep, instead of using `content`, the notification message goes in the `description` of the first rich embed -- both consistent with Discord's webhook format.

# Examples
## Achievement Diary
<details>
<summary>Before</summary>

```json
{
  "type": "ACHIEVEMENT_DIARY",
  "extra": {
	"area": "Desert",
	"difficulty": "HARD",
	"total": 9
  },
  "content": "dankgg has completed the Hard Desert Achievement Diary, for a total of 9 diaries completed",
  "playerName": "dankgg",
  "embeds": []
}
```
</details>

<details>
<summary>After</summary>

```json
{
  "type": "ACHIEVEMENT_DIARY",
  "extra": {
    "area": "Desert",
    "difficulty": "HARD",
    "total": 9
  },
  "playerName": "dankgg",
  "embeds": [
    {
      "title": "Achievement Diary",
      "description": "dankgg has completed the Hard Desert Achievement Diary, for a total of 9 diaries completed",
      "author": {
        "name": "dankgg"
      },
      "color": 15990936,
      "image": {
        "url": "attachment://achievementDiaryImage.png"
      },
      "thumbnail": {
        "url": "https://oldschool.runescape.wiki/images/Achievement_Diaries.png"
      },
      "footer": {
        "text": "Powered by Dink",
        "icon_url": "https://github.com/pajlads/DinkPlugin/raw/master/icon.png"
      },
      "timestamp": "2023-01-02T02:00:00Z"
    }
  ]
}
```
</details>

**Comment by iProdigy** [comment]
Testing is now easy; just set the environmental variable `TEST_WEBHOOK_URL` and run `./gradlew test` (or can do individual tests rather than the full suite)

**Comment by zeroquinc** [comment]
Would it be possible to change this line ourselves?

```
"footer": {
        "text": "Powered by Dink",
        "icon_url": "https://github.com/pajlads/DinkPlugin/raw/master/icon.png"
```

I respect that you guys advertise your own plugin, I was just wondering about this.

Is it possible to change the json file at all?

By the way the embedded looks really great! 👀

**Comment by iProdigy** [comment]
> Would it be possible to change this line ourselves?

Yeah I'll add a config option (e.g., in advanced section) that allows opt-out, or remove it altogether, depending on how the other maintainers feel

File: src/main/java/dinkplugin/message/NotificationBody.java [hunk 4]
```diff
@@ -1,39 +1,46 @@
...
-    }
+    /*
+     * Discord fields
+     */
+    @Nullable
+    @SerializedName("content")
+    private String computedDiscordContent; // this should be set by DiscordMessageHandler, not notifiers
+    private String username;
+    private boolean tts;
+    private @SerializedName("avatar_url") String avatarUrl;
```
Comment (iProdigy) [inline src/main/java/dinkplugin/message/NotificationBody.java:43]: these fields are unused

**Review by iProdigy — COMMENTED** [review]
Should we remove the following webhook fields since they are never written to? Or comment out? Our let them remain (for some future use)?

File: src/main/java/dinkplugin/message/NotificationBody.java [hunk 5]
```diff
@@ -1,39 +1,42 @@
...
 
-    @Builder.Default
-    private List<Embed> embeds = new LinkedList<>();
+    /*
+     * Dink fields
+     */
+    @NotNull
+    NotificationType type;
+    String playerName;
+    AccountType accountType;
```
Comment (iProdigy) [inline src/main/java/dinkplugin/message/NotificationBody.java:27]:
```suggestion
    transient AccountType accountType;
```

This is a new field, if we don't want it sent to the webhook server, can mark it as transient.

File: src/main/java/dinkplugin/util/ColorAdapter.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
+package dinkplugin.util;
+
+import com.google.gson.TypeAdapter;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+
+import java.awt.Color;
+import java.io.IOException;
+
+public class ColorAdapter extends TypeAdapter<Color> {
```
Comment (pajlada) [inline src/main/java/dinkplugin/util/ColorAdapter.java:10]: Can we add a comment here explaining why we need a color adapter? I'm assuming it's a json serialize/deserialize class for the `java.awt.Color` type?

File: src/main/java/dinkplugin/util/InstantAdapter.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+import com.google.gson.TypeAdapter;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.time.format.DateTimeParseException;
+
+public class InstantAdapter extends TypeAdapter<Instant> {
```
Comment (pajlada) [inline src/main/java/dinkplugin/util/InstantAdapter.java:11]: Same thing as with the ColorAdapter

File: src/main/java/dinkplugin/util/Utils.java [hunk 8]
```diff
@@ -18,6 +20,8 @@
 @UtilityClass
 public class Utils {
 
+    public static final String WIKI_IMG_URL = "https://oldschool.runescape.wiki/images/";
```
Comment (pajlada) [inline src/main/java/dinkplugin/util/Utils.java:23]: Maybe rename to specify it's a URL to be used as a base, not by itself. `WIKI_IMG_URL_BASE`?

File: src/main/java/dinkplugin/util/Utils.java [hunk 9]
```diff
@@ -30,6 +34,24 @@ public String getPlayerName(Client client) {
...
+                return "https://oldschool.runescape.wiki/images/Hardcore_ironman_chat_badge.png";
+            case GROUP_IRONMAN:
+                return "https://oldschool.runescape.wiki/images/Group_ironman_chat_badge.png";
+            case HARDCORE_GROUP_IRONMAN:
+                return "https://oldschool.runescape.wiki/images/Hardcore_group_ironman_chat_badge.png";
+            default:
+                return null;
+        }
+    }
+
```
Comment (pajlada) [inline src/main/java/dinkplugin/util/Utils.java:54]: Should we use `WIKI_IMG_URL(_BASE)` here?

File: src/main/java/dinkplugin/message/NotificationBody.java (hunk 5 above)
Comment (pajlada) [reply src/main/java/dinkplugin/message/NotificationBody.java:27]: I think this field is fine to pass along

File: src/main/java/dinkplugin/message/NotificationBody.java [hunk 10]
```diff
@@ -1,39 +1,42 @@
...
 
-    @Builder.Default
-    @EqualsAndHashCode.Exclude
-    private transient String screenshotFile = "image.png";
+    /*
+     * Discord fields
+     */
+    @Nullable
+    @SerializedName("content")
+    String computedDiscordContent; // this should be set by DiscordMessageHandler, not notifiers
```
Comment (pajlada) [inline src/main/java/dinkplugin/message/NotificationBody.java:38]: Specify that this is only set/sent if `richDiscordEmbeds` is disabled?

File: src/main/java/dinkplugin/notifiers/data/LevelNotificationData.java [hunk 11]
```diff
@@ -1,11 +1,30 @@
...
 
 @Value
-public class LevelNotificationData {
+@EqualsAndHashCode(callSuper = false)
+public class LevelNotificationData extends NotificationData {
     Map<String, Integer> levelledSkills;
     Map<String, Integer> allSkills;
+
+    @Override
+    public List<Field> getFields() {
```
Comment (pajlada) [inline src/main/java/dinkplugin/notifiers/data/LevelNotificationData.java:18]: I don't like this field since it looks through all your skills, so if a user levels their agility to 83 but has 103 attack & defense, this will say "Total 99+ skills: 2" which seems irrelevant

File: src/test/java/dinkplugin/notifiers/MockedNotifierTest.java [hunk 12]
```diff
@@ -1,54 +1,91 @@
...
+import static org.mockito.Mockito.doAnswer;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
 
-class MockedNotifierTest extends MockedTestBase {
+abstract class MockedNotifierTest extends MockedTestBase {
 
     protected static final String PLAYER_NAME = "dank";
-    protected static final String PRIMARY_WEBHOOK_URL = "";
+    protected static final String PRIMARY_WEBHOOK_URL = System.getenv("TEST_WEBHOOK_URL");
```
Comment (pajlada) [inline src/test/java/dinkplugin/notifiers/MockedNotifierTest.java:35]: TODO: Figure out where we document this for developers? bottom of readme?

File: src/main/java/dinkplugin/message/Fieldable.java [hunk 13]
```diff
@@ -0,0 +1,7 @@
+package dinkplugin.message;
+
+import java.util.List;
+
+public interface Fieldable {
+    List<Field> getFields();
+}
```
Comment (pajlada) [inline src/main/java/dinkplugin/message/Fieldable.java:7]: Might be Java noob, but the Fieldable interface feels unnecessary to me when we have `NotificationData` as a base class already, since it's a base class that only implements a single interface

File: src/main/java/dinkplugin/message/DiscordMessageHandler.java [hunk 14]
```diff
@@ -107,7 +125,9 @@ private void sendToMultiple(Collection<HttpUrl> urls, MultipartBody.Builder reqB
         urls.forEach(url -> sendMessage(url, reqBodyBuilder, 0));
     }
 
+    @SneakyThrows
```
Comment (pajlada) [inline src/main/java/dinkplugin/message/DiscordMessageHandler.java:128]: Why is this necessary now?

File: src/main/java/dinkplugin/message/DiscordMessageHandler.java [hunk 15]
```diff
@@ -136,13 +156,57 @@ public void onFailure(@NotNull Call call, @NotNull IOException e) {
...
         });
+
+        if (latch != null)
+            latch.await();
+    }
+
+    @VisibleForTesting
+    public boolean isAsync() {
+        return true;
+    }
```
Comment (pajlada) [inline src/main/java/dinkplugin/message/DiscordMessageHandler.java:181]: Why is it necessary for us to be able to control asyncness for the test notifications? does it have to do with rate limits or something?

**Review by pajlada — CHANGES_REQUESTED** [review]
Might be nitpicking, happy to chat

File: src/main/java/dinkplugin/notifiers/data/LevelNotificationData.java (hunk 11 above)
Comment (iProdigy) [reply src/main/java/dinkplugin/notifiers/data/LevelNotificationData.java:18]: It only adds the 99+ field if the *leveled* skill has maxed (e.g., agility hit 99, indicate that user has 3 99's now)

Comment (pajlada) [reply src/main/java/dinkplugin/notifiers/data/LevelNotificationData.java:18]: Ah yeah, the L19 if handles that.
I'm still not sold on the field personally. Would be nice with someone else's opinion too

File: src/main/java/dinkplugin/message/DiscordMessageHandler.java (hunk 14 above)
Comment (iProdigy) [reply src/main/java/dinkplugin/message/DiscordMessageHandler.java:128]: `CountdownLatch#await` throws `InterruptedException`. Since `await` is only called for test cases, I think it's adequate to not add the checked exception to the method declaration of `sendMessage`

File: src/main/java/dinkplugin/message/DiscordMessageHandler.java (hunk 15 above)
Comment (iProdigy) [reply src/main/java/dinkplugin/message/DiscordMessageHandler.java:181]: This was added for `TEST_WEBHOOK_URL` to function; when `sendMessage` is not blocking, junit terminates each test case too early (i.e., before it reliably has a chance to send the message to the webhook server) because the work is delegated to some threadpool. 

I can add a comment explaining this or we can consider alternative ways to get these tests working:
* Add `Thread#sleep` at the bottom of each test (bad imo)
* Switch `DiscordMessageHandler` to use `execute` instead of `enqueue` on okhttp requests, but perform this call on the passed `ScheduledExecutorService`. Then, in tests, we can define a special `Executor` that does the work on the main thread rather than creating a threadpool. This will effectively block the main thread without using `CountdownLatch`

File: src/main/java/dinkplugin/notifiers/data/LevelNotificationData.java (hunk 11 above)
Comment (Felanbird) [reply src/main/java/dinkplugin/notifiers/data/LevelNotificationData.java:18]: I can take it or leave it, I understand the reasoning but it probably makes no sense for it to re-fire on virtual levels.

File: CHANGELOG.md [hunk 16]
```diff
@@ -1,5 +1,7 @@
 ## Unreleased
 
+- Major: Use Discord rich embed format for webhooks. (#110)
```
Comment (iProdigy) [inline CHANGELOG.md:3]:
```suggestion
- Major: Use Discord rich embed format for webhooks. This is technically a possible breaking change for users that target non-Discord webhook servers; these users can revert to the old notification format by disabling the `Use Rich Embeds` setting in the `Advanced` section. (#110)
```

File: src/main/java/dinkplugin/util/InstantAdapter.java [hunk 17]
```diff
@@ -8,6 +8,13 @@
 import java.time.Instant;
 import java.time.format.DateTimeParseException;
 
+/**
+ * Serialized and deserializes {@link Instant} instances
+ * from their equivalent ISO-8601 string representation.
+ * <p>
+ * Discord requires this string format, rather than epoch milliseconds
+ * that the default RuneLite Gson instance would produce.
```
Comment (pajlada) [inline src/main/java/dinkplugin/util/InstantAdapter.java:16]: To be consistent with our own documentation

```suggestion
 * that the default RuneLite GSON instance would produce.
```

File: src/main/java/dinkplugin/notifiers/data/LevelNotificationData.java (hunk 11 above)
Comment (pajlada) [reply src/main/java/dinkplugin/notifiers/data/LevelNotificationData.java:18]: compromise: remove it for now and add in a separate PR to have the discussion there instead?

**Review by pajlada — CHANGES_REQUESTED** [review]
Mini nitpick this time, testing document looks great!

**Review by pajlada — APPROVED** [review]
I'm happy with this change, thank you!

**Review by Mm2PL — APPROVED** [review]
